src/v1/value/value_utils.rs

```diff
@@ -0,0 +1,221 @@
+    }
+
+    /// Lookup the quantity of ada(unit: lovelace).
+    pub fn get_ada_amount(&self) -> BigInt {
+        self.get_token_amount(&CurrencySymbol::Ada, &TokenName::ada())
+    }
+
+    /// Insert a new token into the value, or replace the existing quantity.
+    pub fn insert_token(&self, cs: &CurrencySymbol, tn: &TokenName, a: &BigInt) -> Self {
+        let mut result_map = self.0.clone();
+
+        result_map
+            .entry(cs.clone())
+            .and_modify(|tn_map| {
+                tn_map
+                    .entry(tn.clone())
+                    .and_modify(|old_a| {
+                        *old_a = a.clone();
+                    })
+                    .or_insert_with(|| a.clone());
+            })
+            .or_insert_with(|| singleton((tn.clone(), a.clone())));
+
+        Self(result_map)
+    }
+
+    /// Return true if the value don't have any entries.
+    pub fn is_empty(&self) -> bool {
+        self.0.is_empty()
+    }
+
+    /// Remove all tokens whose quantity is zero.
+    pub fn normalize(&self) -> Self {
+        self.filter(|_, _, a| a.is_zero().not())
+    }
+
+    /// Apply a function to each token of the value, and use its result as the new amount.
+    pub fn map_amount<F>(&self, mut f: F) -> Self
+    where
+        F: FnMut(&CurrencySymbol, &TokenName, &BigInt) -> BigInt,
+    {
+        self.filter_map_amount(|cs, tn, a| Some(f(cs, tn, a)))
+    }
+
+    /// Apply a predicate to tokens.
+    pub fn filter<F>(&self, mut f: F) -> Self
+    where
+        F: FnMut(&CurrencySymbol, &TokenName, &BigInt) -> bool,
+    {
+        self.filter_map_amount(|cs, tn, a| f(cs, tn, a).then(|| a.clone()))
+    }
+
+    /// Apply a function to each token of the value. If the result is None, the token entry will be
+    /// removed.
+    ///
+    /// Note that if the name-quantity map of any given currency symbols is empty, the currency entry
+    /// will be removed from the top-level map entirely.
+    pub fn filter_map_amount<F>(&self, mut f: F) -> Self
+    where
+        F: FnMut(&CurrencySymbol, &TokenName, &BigInt) -> Option<BigInt>,
+    {
+        Value(
+            (&self.0)
+                .into_iter()
+                .filter_map(|(cs, tn_map)| {
+                    let filtered_tn_map = tn_map
+                        .into_iter()
+                        .filter_map(|(tn, a)| f(cs, tn, a).map(|a| (tn.clone(), a)))
+                        .collect::<BTreeMap<TokenName, BigInt>>();
+
+                    if filtered_tn_map.is_empty() {
+                        None
+                    } else {
+                        Some((cs.clone(), filtered_tn_map))
+                    }
+                })
+                .collect(),
+        )
+    }
+}
+
+impl Default for Value {
+    fn default() -> Self {
+        Self(BTreeMap::new())
+    }
+}
+
+impl Zero for Value {
+    fn zero() -> Self {
+        Default::default()
+    }
+
+    fn is_zero(&self) -> bool {
```

Comment:
This could give false negatives, if a Value is not normalized

Reply:
I wonder what laws should hold for this trait. What I had in mind was that `x == zero() && x.is_zero()` should always be true.

Reply:
Ideally we should have types for unsorted and unnormalized variants, just like in plutarch. But that'd introduce extra complexcity, and I'd prefer doing it in a new pr.